otherlibs/systhreads/st_posix.h

```diff
@@ -115,44 +123,76 @@ Caml_inline void st_thread_set_id(intnat id)
    threads. */
 
 typedef struct {
-  pthread_mutex_t lock;         /* to protect contents  */
-  int busy;                     /* 0 = free, 1 = taken */
-  volatile int waiters;         /* number of threads waiting on master lock */
-  pthread_cond_t is_free;       /* signaled when free */
+  pthread_mutex_t lock;           /* to protect contents */
+  uintnat busy;                   /* 0 = free, 1 = taken */
+  atomic_uintnat waiters;         /* number of threads waiting on master lock */
+  pthread_cond_t is_free;         /* signaled when free */
 } st_masterlock;
 
 static void st_masterlock_init(st_masterlock * m)
 {
+
   pthread_mutex_init(&m->lock, NULL);
   pthread_cond_init(&m->is_free, NULL);
   m->busy = 1;
-  m->waiters = 0;
+  atomic_store_rel(&m->waiters, 0);
+
+  return;
+};
+
+static void st_bt_lock_acquire(st_masterlock *m) {
+
+  // We do not want to signal the backup thread is it is not "working"
+  // as it may very well not be, because we could have just resumed
+  // execution from another thread right away.
+  if (caml_bt_is_in_blocking_section()) {
+    caml_bt_enter_ocaml();
+  }
+
+  caml_acquire_domain_lock();
+
+  return;
 }
 
-static void st_masterlock_acquire(st_masterlock * m)
+static void st_bt_lock_release(st_masterlock *m) {
+
+  // Here we do want to signal the backup thread iff there's
```

Comment:
Use `/* */` for comments.

Reply:
Fixed in https://github.com/ocaml-multicore/ocaml/commit/a8d0e5f2dd50cee50c63a80ef7aeab7df20eae32